src/state/internal/createDerivedStore.ts

```diff
@@ -0,0 +1,376 @@
+import { debounce, identity } from 'lodash';
+import { useStoreWithEqualityFn } from 'zustand/traditional';
+import { StoreApi } from 'zustand/vanilla';
+import { IS_DEV } from '@/env';
+import { pluralize } from '@/worklets/strings';
+import { createPathFinder, getOrCreateProxy, PathFinder } from './derivedStore/deriveProxy';
+import {
+  BaseRainbowStore,
+  DebounceOptions,
+  DeriveOptions,
+  DerivedRainbowStore,
+  EqualityFn,
+  Listener,
+  RainbowStore,
+  Selector,
+  SubscribeArgs,
+  UnsubscribeFn,
+  WithFlushUpdates,
+} from './types';
+
+// ============ Store Creator ================================================== //
+
+/**
+ * ### `createDerivedStore`
+ *
+ * Creates a **read-only** store derived from one or more underlying Zustand stores.
+ *
+ * ---
+ * The `deriveFunction` is called whenever its dependencies change, producing a new derived state.
+ * Dependencies are automatically tracked through a special `$` helper, which supports:
+ *
+ * 1) **Selector-based** usage:
+ *    ```ts
+ *    $ => {
+ *      const user = $(useUserStore, s => s.user, shallowEqual);
+ *      const theme = $(useSettingsStore, s => s.appearance.theme);
+ *      return { user, theme, isAdmin: user?.roles.includes('admin') };
+ *    }
+ *    ```
+ *
+ * 2) **Proxy-based** usage (auto-built selectors for nested properties):
+ *    ```ts
+ *    $ => {
+ *      const { user } = $(useUserStore); // Subscribe to `user`
+ *      const theme = $(useSettingsStore).appearance.theme; // Subscribe to `theme`
+ *      return { isAdmin: user?.roles.includes('admin'), theme, user };
+ *    }
+ *    ```
+ *
+ * ---
+ * Derived stores automatically unsubscribe from all dependencies when no consumers remain, and
+ * resubscribe when new consumers appear. The returned function doubles as:
+ *
+ * - A **React hook** (`const state = useDerivedStore(selector, equalityFn?)`)
+ * - A **store object** with `getState()`, `subscribe()`, and `destroy()`
+ *
+ * ---
+ * You can optionally pass a second parameter (either an equality function or a config object)
+ * to enable debouncing, customize the equality function, or set `stableSubscriptions: true`.
+ *
+ * (If stable subscriptions are enabled, dependencies are established once and are not rebuilt
+ * on subsequent re-derives, which can be a performance win for certain workloads.)
```

Comment:
Just to see If I understand - `stableSubscriptions` means it only ever listens to the $ subscriptions from the first time it's ever run?

Reply:
Yeah, if enabled, the store will determine subscriptions once, each time its `watchers` go from 0 → 1, and it'll keep those subscriptions alive through subsequent derivations, until the store is no longer subscribed to.

So if a store goes from inactive → active (0 → 1 subscriptions) → inactive → active, with `stableSubscriptions` enabled that'll result in two subscription rebuilds, no matter how many times your derive function runs.